feat: warn when `--routing-url` is unroutable

trevor-scheer opened the pull request:
Supersedes #1543 (took a different approach, didn't care to destroy / overwrite the history there)

Fixes https://github.com/apollographql/rover/issues/1477

This feature addition adds URL validation to the rover subgraph publish `--routing-url` flag. It also introduces a dependent flag to bypass warnings/prompts `--allow-invalid-routing-url`.

Taken from @EverlastingBugstopper 's https://github.com/apollographql/rover/issues/1477#issuecomment-1397506739:
When a `--routing-url` is passed to rover subgraph publish:

In the case where a human is manually doing a publish (this can be determined by checking for a TTY) then Rover prints a warning that the URL is invalid with a prompt (`[y/N]`) to continue with the publish.

In the case where CI or a script is doing the publish (check for a TTY) then Rover prints a warning saying the routing url is invalid, but still allow the publish to go through. We might also want to make the warning include a notice that this behavior may break in future versions of Rover if the `--allow-invalid-routing-url` flag is not passed.

In both scenarios (and in future versions of Rover), passing the `--allow-invalid-routing-url` flag should disable the warnings entirely and allow the invalid routing url to be used.

When a `--routing-url` is not passed to rover subgraph publish:

Perform a GraphQL query against the platform API to get the routing url for that subgraph
If that URL is invalid, follow the same error/warn procedure as described above

<!--
First, 🌠 thank you 🌠 for taking the time to consider a contribution to Apollo!

Here are some important details to follow:

* ⏰ Your time is important
        To save your precious time, if the contribution you are making will
        take more than an hour, please make sure it has been discussed in an
        issue first. This is especially true for feature requests!

* 💡 Features
        Feature requests can be created and discussed within a GitHub Issue.
        Be sure to search for existing feature requests (and related issues!)
        prior to opening a new request. If an existing issue covers the need,
        please upvote that issue by using the 👍 emote, rather than opening a
        new issue.

* 🕷 Bug fixes
        These can be created and discussed in this repository. When fixing a bug,
        please _try_ to add a test which verifies the fix.  If you cannot, you should
        still submit the PR but we may still ask you (and help you!) to create a test.

* 📖 Contribution guidelines
        Follow https://github.com/apollographql/rover/blob/HEAD/CONTRIBUTING.md
        when submitting a pull request.  Make sure existing tests still pass, and add
        tests for all new behavior.

* ✏️ Explain your pull request
        Describe the big picture of your changes here to communicate to what
        your pull request is meant to accomplish. Provide 🔗 links 🔗 to
        associated issues!

We hope you will find this to be a positive experience! Open source
contribution can be intimidating and we hope to alleviate that pain as much
as possible. Without following these guidelines, you may be missing context
that can help you succeed with your contribution, which is why we encourage
discussion first. Ultimately, there is no guarantee that we will be able to
merge your pull-request, but by following these guidelines we can try to
avoid disappointment.

-->

trevor-scheer committed:
checkpoint

trevor-scheer committed:
checkpoint

trevor-scheer committed:
Simplify and handle different types of parse errors

trevor-scheer commented on src/command/subgraph/publish.rs:
```diff
@@ -74,4 +112,186 @@ impl Publish {
+                Err(anyhow!("Publish cancelled by user").into())
+            }
+        }
+    }
```
@EverlastingBugstopper this fn is fairly duplicative of the prompt default no util, but I did want the ability to stub the reader/writer. Happy to port this change to the util and update usages (I think there's only 1).

EverlastingBugstopper replied:
porting would be great! would you mind doing so in a follow-up after we merge?

trevor-scheer replied:
Follow-up sounds great!

trevor-scheer commented on src/command/subgraph/publish.rs:
```diff
@@ -74,4 +112,186 @@ impl Publish {
+    }
+
+    pub fn issue_warnings_non_tty(
+        // a None value here means we're not in a tty, so we can't prompt
```
Note to self: this changed from `Option<bool>` and comment needs update

trevor-scheer commented on src/command/subgraph/publish.rs:
```diff
@@ -74,4 +112,186 @@ impl Publish {
+        // a None value here means we're not in a tty, so we can't prompt
+        is_atty: bool,
+        // for testing purposes, so we can inject an output stream
+        output: &mut dyn std::io::Write,
```
Align on calling this `writer` everywhere (open to input on the name, but that's what I went with elsewhere)

trevor-scheer committed:
align on reader/writer naming

trevor-scheer committed:
unwind / simplify is_atty logic

trevor-scheer commented on src/command/subgraph/publish.rs:
```diff
@@ -74,4 +112,183 @@ impl Publish {
+                            )?;
+                        } else {
+                            Self::issue_non_tty_warnings(writer)?;
+                        }
```
Pattern repeated 3x with only difference being the prompt message

EverlastingBugstopper committed:
Merge branch 'main' into trevor/invalid-subgraph-routing-url

EverlastingBugstopper commented on crates/rover-client/src/operations/subgraph/mod.rs:
```diff
@@ -13,6 +13,9 @@ pub mod fetch;
 /// "subgraph publish" command execution
 pub mod publish;
 
+/// "subgraph publish" no (--routing-url) command execution
```
```suggestion
/// query for a single subgraph's routing URL
```

EverlastingBugstopper commented on crates/rover-client/src/operations/subgraph/routing_url/runner.rs:
```diff
@@ -0,0 +1,116 @@
+pub fn run(
+    input: SubgraphRoutingUrlInput,
+    client: &StudioClient,
+) -> Result<Option<String>, RoverClientError> {
```
```suggestion
) -> Result<String, RoverClientError> {
```

I think this should be a `String` and not an `Option<String>` because a subgraph _must_ have a routing URL if it exists in GraphOS. Therefore, if one _doesn't_ exist, that is an error (probably subgraph doesn't exist or malformed response)

trevor-scheer replied:
👍 35d6dc4

EverlastingBugstopper commented on crates/rover-client/src/operations/subgraph/routing_url/runner.rs:
```diff
@@ -0,0 +1,116 @@
+        match maybe_variant {
+            SubgraphRoutingUrlGraphVariant::GraphVariant(variant) => {
+                if let Some(subgraph) = variant.subgraph {
+                    Ok(subgraph.url)
```
if this is a `String` instead of `Option<String>` we'll need something like 
```suggestion
                    if let Some(url) = subgraph.url {
                      Ok(url)
                    } else {
                      Err(RoverClientError::MalformedResponse { null_field: "graph.variant.subgraph.url".to_string())
                    }
```

trevor-scheer replied:
35d6dc4

EverlastingBugstopper commented on crates/rover-client/src/operations/subgraph/routing_url/runner.rs:
```diff
@@ -0,0 +1,116 @@
+                if let Some(subgraph) = variant.subgraph {
+                    Ok(subgraph.url)
+                } else {
+                    Err(RoverClientError::ExpectedFederatedGraph {
```
not sure if this is the right error code here - if you run this for a subgraph that doesn't exist on an existing federated graph you'll get a confusing error message

trevor-scheer replied:
Ah, good catch. So do we actually have 2 possible error conditions here with no way to differentiate based on the response?
* You ran `rover subgraph publish` with no `--routing-url` (so we ran this query) against a graph that isn't federated
* You ran `rover subgraph publish` with no `--routing-url` (so we ran this query) for a subgraph which hasn't been published yet

If so, do you have any ideas? Or do we respond with an error that says "either this or that happened"?

EverlastingBugstopper replied:
I think we're covered on the non-federated one because we run `IsFederatedQuery` beforehand which is good. 

I think the two cases we need to differentiate between here are 

- You ran `rover subgraph publish` with no `--routing-url` (so we ran this query) for a subgraph which hasn't been published yet
  - This should return `null` for `graph.variant.subgraph` and we should error with "this subgraph doesn't exist yet and you need a URL to publish a subgraph for the first time" this error already gets returned from studio, but we can now detect it ourselves (which is good actually because running that kind of publish leaves things in a weird state sometimes)
- You ran `rover subgraph publish` with no `--routing-url` (so we ran this query) for a subgraph which does exist, but for some reason our server implementation returns `null` for `graph.variant.subgraph.routing_url`
  - this theoretically shouldn't happen, but because literally any field can be `null` at any time for any reason, we need to handle this in our client code. in this case we have a special `MalformedResponse` variant on `RoverClientError`

trevor-scheer replied:
[010f1ed](https://github.com/apollographql/rover/pull/1565/commits/010f1ed310e83849772002a245f4011b715dd66d)

EverlastingBugstopper commented on src/command/subgraph/publish.rs:
```diff
@@ -74,4 +112,183 @@ impl Publish {
         })
     }
+
+    fn handle_maybe_invalid_routing_url(
```
let's convert this from an associated function to a method on the struct
```suggestion
    fn handle_maybe_invalid_routing_url(&self
```

then we can use `&self.routing_url` instead of needing to pass `self.routing_url.clone()`

trevor-scheer replied:
Discussed offline - having this fn not require `self` is great for testing purposes (so we're now using a reference to the `routing_url` instead of cloning it).
35d6dc4

EverlastingBugstopper commented on src/command/subgraph/publish.rs:
```diff
@@ -74,4 +112,183 @@ impl Publish {
+                    if !vec!["http", "https"].contains(&parsed_url.scheme()) {
+                        if is_atty {
+                            Self::prompt_for_publish(
+                                format!("The `{}` protocol is not supported by router, it expects either `http` or `https`. Continuing the publish will make this subgraph unreachable by your supergraph. Would you still like to publish?", &parsed_url.scheme()).as_str(),
```
```suggestion
                                format!("The `{}` protocol is not supported by the router. Valid protocols are `http` and `https`. Continuing the publish will make this subgraph unreachable by your supergraph. Would you still like to publish?", &parsed_url.scheme()).as_str(),
```

EverlastingBugstopper commented on src/command/subgraph/publish.rs:
```diff
@@ -74,4 +112,183 @@ impl Publish {
+        if std::str::from_utf8(&response).unwrap().to_lowercase() == *"y" {
+            Ok(Some(true))
+        } else {
+            Err(anyhow!("Publish cancelled by user").into())
```
```suggestion
            Err(anyhow!("You cancelled the publish.").into())
```

EverlastingBugstopper reviewed (changes requested):
very close, just a few more small comments

trevor-scheer committed:
Update crates/rover-client/src/operations/subgraph/mod.rs

Co-authored-by: Avery Harnish <avery@apollographql.com>

trevor-scheer committed:
Update src/command/subgraph/publish.rs

Co-authored-by: Avery Harnish <avery@apollographql.com>

trevor-scheer committed:
Update src/command/subgraph/publish.rs

Co-authored-by: Avery Harnish <avery@apollographql.com>

trevor-scheer committed:
update tests based on copy updates

trevor-scheer committed:
address feedback (with one TODO remaining)

trevor-scheer committed:
lint

trevor-scheer committed:
Merge branch 'main' into trevor/invalid-subgraph-routing-url

trevor-scheer committed:
Fix error handling for unpublished case

EverlastingBugstopper committed:
Merge branch 'main' into trevor/invalid-subgraph-routing-url

EverlastingBugstopper committed:
fix: removes duplicated `IsFederatedQuery` and allows `--convert` (#1575)

This PR removes the `IsFederatedQuery` check from the routing URL
fetcher. This enables the flow where you are trying to publish a
subgraph for the first time to a variant that used to be a monograph.
Additionally, it removes the need for the `IsFederatedQuery` to be sent
twice on each `rover subgraph publish`.
